include/ck/utility/transpose_vectors.hpp

```diff
@@ -83,5 +91,174 @@ struct transpose_vectors<half_t, NX, NY>
+
+    float v0 = static_cast<float>(vx.template AsType<half_t>()[I0]);
+    float v1 = static_cast<float>(vx.template AsType<half_t>()[I1]);
+
+    vy.template AsType<bhalf_t>()(I0) = ck::type_convert<bhalf_t>(v0);
+    vy.template AsType<bhalf_t>()(I1) = ck::type_convert<bhalf_t>(v1);
+
+    y = vy.template AsType<bhalf2_t>()[I0];
+#else
+    float y0{0}, y1{0};
+    asm volatile("\n \
+            v_cvt_f32_f16 %0, %1 \n \
+            "
+                 : "=v"(y0)
+                 : "v"(x));
+    asm volatile("\n \
+            v_cvt_f32_f16 %0, %1 src0_sel:WORD_1\n \
+            "
+                 : "=v"(y1)
+                 : "v"(x));
+    asm volatile("\n \
+            v_pack_b32_f16 %0, %1, %2 op_sel:[1, 1] \n \
+            "
+                 : "=v"(y)
+                 : "v"(y0), "v"(y1));
+#endif
+}
+__device__ void
+transpose_half_to_bhalf_2x2(const half2_t& x0, const half2_t& x1, bhalf2_t& y0, bhalf2_t& y1)
+{
+#if 0
+    static constexpr auto I0 = Number<0>{};
+    static constexpr auto I1 = Number<1>{};
+
+    const vector_type<half_t, 2> vx0{x0}, vx1{x1};
+    vector_type<bhalf_t, 2> vy0, vy1;
+
+    float v0                           = static_cast<float>(vx0.template AsType<half_t>()[I0]);
+    float v1                           = static_cast<float>(vx1.template AsType<half_t>()[I0]);
+    vy0.template AsType<bhalf_t>()(I0) = ck::type_convert<bhalf_t>(v0);
+    vy0.template AsType<bhalf_t>()(I1) = ck::type_convert<bhalf_t>(v1);
+
+    v0                                 = static_cast<float>(vx0.template AsType<half_t>()[I1]);
+    v1                                 = static_cast<float>(vx1.template AsType<half_t>()[I1]);
+    vy1.template AsType<bhalf_t>()(I0) = ck::type_convert<bhalf_t>(v0);
+    vy1.template AsType<bhalf_t>()(I1) = ck::type_convert<bhalf_t>(v1);
+
+    y0 = vy0.template AsType<bhalf2_t>()[I0];
+    y1 = vy1.template AsType<bhalf2_t>()[I0];
+#else
+    float yv0{0}, yv1{0};
+    asm volatile("\n \
+            v_cvt_f32_f16 %0, %1 \n \
+            "
+                 : "=v"(yv0)
+                 : "v"(x0));
+
+    asm volatile("\n \
+            v_cvt_f32_f16 %0, %1 \n \
+            "
+                 : "=v"(yv1)
+                 : "v"(x1));
+
+    asm volatile("\n \
+            v_pack_b32_f16 %0, %1, %2 op_sel:[1, 1] \n \
+            "
+                 : "=v"(y0)
+                 : "v"(yv0), "v"(yv1));
+
+    asm volatile("\n \
+            v_cvt_f32_f16 %0, %1 src0_sel:WORD_1\n \
+            "
+                 : "=v"(yv0)
+                 : "v"(x0));
+    asm volatile("\n \
+            v_cvt_f32_f16 %0, %1 src0_sel:WORD_1\n \
+            "
+                 : "=v"(yv1)
+                 : "v"(x1));
+
+    asm volatile("\n \
+            v_pack_b32_f16 %0, %1, %2 op_sel:[1, 1] \n \
+            "
+                 : "=v"(y1)
+                 : "v"(yv0), "v"(yv1));
+#endif
+}
+
+template <index_t NX, index_t NY>
+struct transpose_convert_vectors<half_t, half_t, NX, NY>
```

Comment:
Similar to above comments, if we simply do type conversion after the optional transpose, then we probably won't need these additional transpose & convert just for the fp16 denorm special case. I am not sure of the performance impacts doing so, though.